cartographer_grpc/metrics/prometheus/metrics_test.cc

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/metrics/family_factory.h"
+#include "cartographer/metrics/register.h"
+#include "cartographer_grpc/metrics/prometheus/family_factory.h"
+#include "glog/logging.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "prometheus/exposer.h"
+
+namespace cartographer_grpc {
+namespace metrics {
+namespace prometheus {
+namespace {
+
+static auto* kCounter = cartographer::metrics::Counter::Null();
+static auto* kGauge = cartographer::metrics::Gauge::Null();
+static auto* kScoresMetric = cartographer::metrics::Histogram::Null();
+
+const char kLabelKey[] = "kind";
+const char kLabelValue[] = "score";
+const std::array<double, 5> kObserveScores = {{-1, 0.11, 0.2, 0.5, 2}};
+
+class Algorithm {
+ public:
+  static void RegisterMetrics(cartographer::metrics::FamilyFactory* factory) {
+    auto boundaries = cartographer::metrics::Histogram::FixedWidth(0.05, 20);
+    auto* scores_family = factory->NewHistogramFamily(
+        "/algorithm/scores", "Scores achieved", boundaries);
+    kScoresMetric = scores_family->Add({{kLabelKey, kLabelValue}});
+  }
+  void Run() {
+    for (double score : kObserveScores) {
+      kScoresMetric->Observe(score);
+    }
+  }
+};
+
+TEST(MetricsTest, CollectCounter) {
+  FamilyFactory factory;
+  auto* counter_family = factory.NewCounterFamily("/test/hits", "Hits");
+  kCounter = counter_family->Add({{kLabelKey, kLabelValue}});
+  kCounter->Increment();
+  kCounter->Increment(5);
+  double expected_value = 1 + 5;
+  std::vector<::io::prometheus::client::MetricFamily> collected;
+  {
+    std::shared_ptr<::prometheus::Collectable> collectable;
+    CHECK(collectable = factory.GetCollectable().lock());
+    collected = collectable->Collect();
+  }
+  ASSERT_EQ(collected.size(), 1);
+  ASSERT_EQ(collected[0].metric_size(), 1);
+  EXPECT_THAT(
+      collected[0].metric(0).label(),
+      testing::AllOf(
+          testing::ElementsAre(testing::Property(
+              &io::prometheus::client::LabelPair::name, kLabelKey)),
+          testing::ElementsAre(testing::Property(
+              &io::prometheus::client::LabelPair::value, kLabelValue))));
+  EXPECT_THAT(collected[0].metric(0).counter().value(),
+              testing::Eq(expected_value));
```

Comment:
Please make sure that testing::Eq is safe for comparing doubles (has an epsilon for equality), or tests will become flaky.

Reply:
Done. Thanks for the fast review.